pkg/kubelet/prober/worker.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prober
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/probe"
+	"k8s.io/kubernetes/pkg/types"
+)
+
+// worker handles the periodic probing of its assigned container. Each worker has a go-routine
+// associated with it which runs the probe loop until the container permanently terminates, or the
+// stop channel is closed. The worker uses the probe Manager's statusManager to get up-to-date
+// container IDs.
+// TODO: Handle liveness probing
+type worker struct {
+	// Channel for stopping the probe, it should be closed to trigger a stop.
+	stop chan struct{}
+
+	// The pod containing this probe (read-only)
+	pod *api.Pod
+
+	// The container to probe (read-only)
+	container api.Container
+
+	// Describes the probe configuration (read-only)
+	spec *api.Probe
+
+	// The last known container ID for this worker.
+	containerID types.UID
+}
+
+// Creates and starts a new probe worker.
+func (m *manager) newWorker(
+	pod *api.Pod,
+	container api.Container) *worker {
+
+	w := &worker{
+		stop:      make(chan struct{}),
+		pod:       pod,
+		container: container,
+		spec:      container.ReadinessProbe,
+	}
+
+	// Start the worker thread.
+	go run(m, w)
+
+	return w
+}
+
+// run periodically probes the container.
+func run(m *manager, w *worker) {
+	probeTicker := time.Tick(m.defaultProbePeriod)
+
+probeLoop:
+	for doProbe(m, w) {
+		// Wait for next probe tick.
+		select {
+		case <-w.stop:
+			break probeLoop
+		case <-probeTicker:
+			// continue
+		}
+	}
+
+	// Clean up.
+	if w.containerID != "" {
+		m.readinessCache.removeReadiness(string(w.containerID))
+	}
+
+	m.workerLock.Lock()
+	defer m.workerLock.Unlock()
+	delete(m.readinessProbes, containerPath{w.pod.UID, w.container.Name})
```

Comment:
Do you want to avoid blocking here, or just avoid referencing the lock from the worker file? I think this is the right place to actually block and delete the worker, but I'm not against moving this code into a function on the manager.

I debated with myself whether run should be `m.run(w)` or `w.run(m)`, and decided since it was ambiguous `run(m, w)` was the way to go.